runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10283,49 +10262,108 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
-
-         iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, doneLabel, iCursor);
+         if(isVariableLen || isLargeConstantLen)
+            {
+            TR::Register * addressReg = srm->findOrCreateScratchRegister();
+            iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
 
-         iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 0, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+            // Repurpose sizeReg to keep loop iterations.
+            if (TR::comp()->target().is64Bit())
+               {
+               iCursor = generateRSInstruction(cg, TR::InstOpCode::SRAG, node, sizeReg, tmp, 8, iCursor);
+               }
+            else
+               {
+               if (comp->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z196))
+                  {
+                  iCursor = generateRSInstruction(cg, TR::InstOpCode::SRAK, node, sizeReg, tmp, 8, iCursor);
+                  }
+               else
+                  {
+                  iCursor = generateRRInstruction(cg, TR::InstOpCode::LR, node, sizeReg, tmp, iCursor);
+                  iCursor = generateRSInstruction(cg, TR::InstOpCode::SRA, node, sizeReg, 8, iCursor);
+                  }
+               }
 
-         // minus 1 from lengthreg since xc auto adds 1 to it
+            // Zero blocks of 256 bytes if size > 256.
+            if(isVariableLen)
+               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, fillerRemLabel, iCursor);
 
-         iCursor = generateEXDispatch(node, cg, lengthReg, shiftReg, iCursor);
-         iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, doneLabel);
-         }
-      cg->stopUsingRegister(addressReg);
-      cg->stopUsingRegister(shiftReg);
-      cg->stopUsingRegister(lengthReg);
+            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerLoopLabel);
+            iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+            iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, sizeReg, fillerLoopLabel);
 
-      if (zeroReg != NULL)
-         {
-         iCursor = generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, zeroReg, zeroReg, iCursor);
+            // Zero the remainder.
+            if(isLargeConstantLen)
+               {
+               // The remainder is a constant value between 1 and 256 for large constant length objects and displacement is 0.
+               allocSize = ((allocSize - 1) % 256) + 1;
```

Comment:
I do not understand this. If we have for example, 4096 bytes to zero, shouldn't we deal whole object inside the fillerLoop, and in that case we should not have any residue left. So this check, would essentially copy the extra 256 bytes, wouldn't it?

Reply:
I change this